torch/onnx/_internal/exporter.py

```diff
@@ -410,7 +439,137 @@ def save(
+        onnx_model = onnxscript_graph.to_model_proto(options.opset_version)
+        return torch.onnx.ExportOutput(
+            onnx_model, self.input_adapter, self.output_adapter
+        )
+
+    def adapt_input(
+        self,
+        adapt_step_cls: Type[InputAdaptStep],
+        model_args: Sequence[Any],
+        model_kwargs: Mapping[str, Any],
+        step_init_args: Optional[Sequence[Any]] = None,
+    ) -> Tuple[Sequence[Any], Mapping[str, Any]]:
+        """Apply an input adapt step to the model args and kwargs.
+        An input adapt step object is initialized, applied and recorded as part of
+        ``self.input_adapter`.
+        Args:
+            adapt_step_cls: The input adapt step class.
+            model_args: The model args.
+            model_kwargs: The model kwargs.
+            step_init_args: The input adapt step initialization arguments.
+        Returns:
+            The adapted model args and kwargs.
+        """
+        step_init_args = step_init_args or ()
+        adapt_step = adapt_step_cls(*step_init_args)
+        self.input_adapter.append_step(adapt_step)
+        return adapt_step.apply(model_args, model_kwargs)
+
+    def adapt_output(
+        self,
+        adapt_step_cls: Type[OutputAdaptStep],
+        model_outputs: Any,
+        step_init_args: Optional[Sequence[Any]] = None,
+    ) -> Any:
+        """Apply an output adapt step to the model outputs.
+        An output adapt step object is initialized, applied and recorded as part of
+        ``self._output_adapter`.
+        Args:
+            adapt_step_cls: The output adapt step class.
+            model_outputs: The model outputs.
+            step_init_args: The input adapt step initialization arguments.
+        Returns:
+            The adapted model outputs.
+        """
+        step_init_args = step_init_args or ()
+        adapt_step = adapt_step_cls(*step_init_args)
+        self.output_adapter.append_step(adapt_step)
+        return adapt_step.apply(model_outputs)
+
+    @abc.abstractmethod
+    def generate_fx(
+        self,
+        options: ResolvedExportOptions,
+        model: Union[torch.nn.Module, Callable],
+        model_args: Sequence[Any],
+        model_kwargs: Mapping[str, Any],
+    ) -> Tuple[torch.fx.GraphModule, Tuple[Any]]:
+        """Analyzes user ``model`` and generates a FX graph.
+        Args:
+            options: The export options.
+            model: The user model.
+            model_args: The model's positional input arguments.
+            model_kwargs: The model's keyword input arguments.
+        Returns:
+            The generated FX Graph, and the model's adapted input arguments.
+        """
+        # By design, only torch.fx.GraphModule is needed
+        # But FXSymbolicTracer modifies model data, which will be needed
+        # to produce the ONNX proto in the next layer.
+        # TODO: Refactor after https://github.com/pytorch/pytorch/pull/98421
+        pass
```

Comment:
nit: abc style
```suggestion
        ...
```

Reply:
Pytorch uses Google style, which uses `pass`?

Reply:
Looks like `...` is what python uses nowadays (I didn't know either): https://docs.python.org/3/library/abc.html#abc.abstractmethod

Reply:
I personally suggest `raise NotImplementedError()` because otherwise subclasses can still call `super().xxx()` and silently do nothing